### PR TITLE
Fix a bug with dpkg lock prevents apt commands run

### PR DESCRIPTION
**Pull request by ufukty**
Related with the issue https://github.com/angristan/openvpn-install/issues/720

### CODE REVIEW

**Comment by angristan** [comment]
Maybe `apt ... || exit 1` would be better ? APT could fail for a number of reasons

**Comment by ufukty** [comment]
This problem occurs nearly every time when executing the script in a machine just launched. And this makes harder to use this script with Terraform and other orchestration/provisioning tools. Because those tools execute tasks immediately. Although you can say it’s those tools’ job to make sure the machine is ready for using apt before launching script.

**Comment by angristan** [comment]
As said earlier exiting the script on APT fail would be great, but going into an infinite loop on APT fail doesn't seem like a good design to me.

I've never encountered the issue, so this has to do with your provider, and more specifically the image you're using. I would suggest using a custom, up-to-date image that doesn't auto-update during the first boot. This has to be handled on your side, not by the script. 🙂

**Comment by randshell** [comment]
@angristan as @hellresistor said https://github.com/angristan/openvpn-install/issues/720#issuecomment-683961800, it happens by default on Ubuntu and possibly on Debian if `unattanded-upgrades` is configured to run at that time.

**Comment by nevesnunes** [comment]
> As said earlier exiting the script on APT fail would be great, but going into an infinite loop on APT fail doesn't seem like a good design to me.
> 
> I've never encountered the issue, so this has to do with your provider, and more specifically the image you're using. I would suggest using a custom, up-to-date image that doesn't auto-update during the first boot. This has to be handled on your side, not by the script. slightly_smiling_face

I was also hit with this issue and see the benefits of having the script fail early, so that the root cause is more evident to the user: e.g. `/etc/openvpn/tls-crypt.key: No such file or directory` does not directly point to issues installing `openvpn`.

We can think of alternatives for the infinite loop:

1. Loop with a fixed number of max attempts, e.g.
```bash
attempts=10
while [ $attempts -gt 0 ]; do
  apt-get install -y openvpn && break
  attempts=$((attempts - 1))
done
```
2. Run the command once, and do some post-install validations, e.g. `command -v openvpn`.

**Comment by ufukty** [comment]
This is not about the OS image. All linux dists have cron tasks that may run any time and sometimes they can conflict with each other. There is undeniable possibility that apt is already running before user fire up the script, even if the system is up-to-date, even it is a custom image or the default one.

The problem is, in the case of failure happened, inexperienced users may not understand why it is not working. A notice about the problem should displayed to user. 

Solution of @nevesnunes is good although it lacks the 1 second of sleep which may do the magic. And since it is 5 lines and repeats at multiple places of script, it could be organized within a function. 

In summary:

```sh
function apt_get_with_retry {
  attempts=0
  until apt-get $@; do 
    sleep 1
    attempts=$((attempts + 1))
    if [[ $attempts -gt 10 ]]; then
      echo "apt-get is busy right now, please try again later"
      exit 1
    fi
  done
}
```

Calls

```sh
apt_get_with_retry install -y openvpn iptables openssl wget ca-certificates curl
# ...
apt_get_with_retry install -y ca-certificates gnupg
# ...
apt_get_with_retry autoremove --purge -y unbound
```

I didn't check if it's working tho. 😎

**Comment by randshell** [comment]
> There is undeniable possibility that apt is already running before user fire up the script, even if the system is up-to-date, even it is a custom image or the default one.

Looks like some real examples are `apt-daily.service` and `apt-daily-upgrade.service`. I wonder why they are not better configured to maybe run at night or at a custom time like `unattanded-upgrades`.

`apt-daily.service` in particular runs at random times
```
OnCalendar=*-*-* 6,18:00
RandomizedDelaySec=12h
```
but `apt-daily-upgrade.service` only between 6 and 7 am
```
OnCalendar=*-*-* 6:00
RandomizedDelaySec=60m
```
@ufukty when did you experienced this issue, not long after reboot or during the day? I think on Ubuntu it's known to wait for some time after boot for the lock file.
> The problem is, in the case of failure happened, inexperienced users may not understand why it is not working. 

Funny enough, the automatic upgrade on Debian based distros is made for inexperienced users. I don't remember anything similar on others.

**Comment by ufukty** [comment]
@randomshell That was right after reboot and recurred for multiple times. I don't think there is significant waiting between booting the system and running the script because of automation. 

Seems like they rolled a new version of apt which introduces a `wait for lock` feature on 7 March 2020. But they only mention `apt` not `apt-get`.

```
* apt(8): Wait for lock (Closes: #754103)
```
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=754103
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=850172

**Comment by randshell** [comment]
I like more the idea of using `apt`, it's the recommended command since Ubuntu 16.04 too.
